examples/connext_dds/lbediscovery/c++11/ShapeType_subscriber.cxx

```diff
@@ -0,0 +1,125 @@
+/*
+ * (c) Copyright, Real-Time Innovations, 2022.  All rights reserved.
+ * RTI grants Licensee a license to use, modify, compile, and create derivative
+ * works of the software solely for use with RTI Connext DDS. Licensee may
+ * redistribute copies of the software provided that all such copies are subject
+ * to this license. The software is provided "as is", with no warranty of any
+ * type, including any warranty for fitness for any purpose. RTI is under no
+ * obligation to maintain or support the software. RTI shall not be liable for
+ * any incidental or consequential damages arising out of the use or inability
+ * to use the software.
+ */
+
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+#include <dds/sub/ddssub.hpp>
+#include <dds/core/ddscore.hpp>
+#include <rti/config/Logger.hpp>  // for logging
+#include "ShapeType.hpp"
+#include "application.hpp"  // for command line parsing and ctrl-c
+
+int process_data(
+        dds::sub::DataReader<ShapeType> reader,
+        std::string reader_name)
+{
+    // Take all samples
+    int count = 0;
+    dds::sub::LoanedSamples<ShapeType> samples = reader.take();
+    for (const auto &sample : samples) {
+        if (sample.info().valid()) {
+            count++;
+            std::cout << reader_name << ": " << sample.data() << std::endl;
+        }
+    }
+
+    return count;
+}  // The LoanedSamples destructor returns the loan
+
+void run_subscriber_application(
+        unsigned int domain_id,
+        unsigned int sample_count)
+{
+    rti::domain::register_type<ShapeType>("ShapeType");
+
+    // IMPORTANT: If you plan to use a domain_id different than 0, remember
+    // to change the value of the domain_id attribute of the domain tag in
+    // the USER_QOS_PROFILES.xml
+    rti::domain::DomainParticipantConfigParams params(domain_id);
+
+    auto default_provider = dds::core::QosProvider::Default();
+
+    // Create the participant and all its entities from the XML file
+    // (USER_QOS_PROFILES.xml)
+    dds::domain::DomainParticipant participant =
+            default_provider.extensions().create_participant_from_config(
+                    "ShapeType_DomainParticipantLibrary::"
+                    "ShapeTypeSubscriberParticipant",
+                    params);
+
+    const std::string reader_names[] = { "Sub::RED_DR",
+                                         "Sub::RED_DR#1",
+                                         "Sub::BLUE_DR",
+                                         "Sub::BLUE_DR#1" };
+    std::vector<dds::sub::DataReader<ShapeType>> readers;
+    std::vector<dds::core::cond::WaitSet> waitsets(4);
+    std::vector<unsigned int> samples_read(4);
+
+    // Retrieve the four DataReaders and configure their
+    // ReadConditions and Waitsets
+    for (int i = 0; i < 4; i++) {
+        readers.push_back(rti::sub::find_datareader_by_name<
```

Comment:
Get the reader first, check for `dds::core::null` (and throw runtime_error), then add it.